src/kernel/chain.cpp

```diff
@@ -16,9 +16,8 @@ interfaces::BlockInfo MakeBlockInfo(const CBlockIndex* index, const CBlock* data
     if (index) {
         info.prev_hash = index->pprev ? index->pprev->phashBlock : nullptr;
         info.height = index->nHeight;
-        LOCK(::cs_main);
-        info.file_number = index->nFile;
-        info.data_pos = index->nDataPos;
+        info.file_number = index->GetFileNum();
+        info.data_pos = index->GetDataPos();
```

Comment:
Shouldn't these be fetched atomically? Maybe there should be a getter that returns all three values, similar to how there's a setter (`CBlockIndex::SetFileData`) than sets all three atomically.

Reply:
Starting to feel like a struct for those 3 members would be appropriate

Reply:
> Shouldn't these be fetched atomically? Maybe there should be a getter that returns all three values, similar to how there's a setter (`CBlockIndex::SetFileData`) than sets all three atomically.

@LarryRuane I didn't add it merely to not introduce that many changes but could be done too. The members set is guarded by `cs_main` (for now) so there is no possible race.

> Starting to feel like a struct for those 3 members would be appropriate

Why exactly?. So far in the code, we either want the block data position or the block undo data position, we don't get both at the same time because they represent different files on disk (thus why they are returned inside `FlatFilePos`, the struct groups only one of them with the file number).

In other words, nData + nFile maps to a "blk<nFile>.dat", while nUndoData + nFile maps to an specific "rev<nFile>.dat".

Reply:
Are you sure these are protected by `cs_main`, is it? Running `gdb test_bitcoin` on the PR branch (leaving some irrelevant stuff out):
```
(gdb) b MakeBlockInfo
(gdb) run
Thread 29 "b-basic block f" hit Breakpoint 1, kernel::MakeBlockInfo (index=0x55555749b018, data=0x0) at kernel/chain.cpp:14
#0  kernel::MakeBlockInfo (index=0x55555749b018, data=0x0) at kernel/chain.cpp:14
#1  0x000055555623fa81 in BaseIndex::ThreadSync (this=0x7fffffffad00) at index/base.cpp:201
(gdb) p cs_main
$1 = {
  <std::recursive_mutex> = {
    <std::__recursive_mutex_base> = {
      _M_mutex = {
        __data = {
          __lock = 0,
          __count = 0,
          __owner = 0,
          }
        }
      }
    }
  }
```
I agree with you that a new struct should just be `nFile` and ` nDataPos` (or `nPos`). This struct can be used for either block data or rev data.

Reply:
> Are you sure these are protected by cs_main, is it?

The members' write process, not the read. If the read would be protected by `cs_main`, this PR wouldn't make any sense :p.
The two places are: the `CBlockIndex` creation time and `CBlockIndex` pruning time.

Still, those `cs_main` locks can also be removed in a follow-up PR. I haven't removed them here because they require further changes.

Not sure if I'm understanding your concern here. You are thinking on a race in-between `GetFileNum` and `GetDataPos`?
Because, if that is the concern, the error that could arise from it is harmless. The read block call would just return false, and callers would continue. But, to not even have to think about such error, we could just call `GetFilePos()` instead of the two getters.

> I agree with you that a new struct should just be nFile and  nDataPos (or nPos). This struct can be used for either block data or rev data.

There is already a method for that: `CBlockIndex::GetFilePos`.